internal/guest/network/netns_test.go

```diff
@@ -0,0 +1,228 @@
+	return func(link netlink.Link, addr *netlink.Addr) error {
+		if addr.IP.String() != expectedIP {
+			return fmt.Errorf("expected to add address %s, instead got %s", expectedIP, addr.IP.String())
+		}
+		expectedMask := net.CIDRMask(prefixLen, totalMaskSize)
+		if !bytes.Equal(addr.Mask, expectedMask) {
+			return fmt.Errorf("expected mask to be %s, instead got %s", expectedMask, addr.Mask)
+		}
+		return nil
+	}
+}
+
+func standardNetlinkRouteAdd(gatewayIP string, table, metric int) func(_ *netlink.Route) error {
+	return func(route *netlink.Route) error {
+		if route.Gw.String() != gatewayIP {
+			return fmt.Errorf("expected to add gateway %s, instead got %s", gatewayIP, route.Gw.String())
+		}
+		if route.Table != table {
+			return fmt.Errorf("expected to use table %d, instead got %d", table, route.Table)
+		}
+		if route.Priority != metric {
+			return fmt.Errorf("expected to use metric %d, instead used %d", metric, route.Priority)
+		}
+		return nil
+	}
+}
+
+type assignIPToLinkTest struct {
+	name          string
+	ifStr         string
+	allocatedIP   string
+	gatewayIP     string
+	prefixLen     uint8
+	totalMaskSize int
+}
+
+var defaultAssignIPToLinkTests = []assignIPToLinkTest{
+	{
+		name:          "ipv4 standard",
+		ifStr:         "eth0",
+		allocatedIP:   "192.168.0.5",
+		gatewayIP:     "192.168.0.100",
+		prefixLen:     uint8(24),
+		totalMaskSize: 32,
+	},
+	{
+		name:          "ipv6 standard",
+		ifStr:         "eth0",
+		allocatedIP:   "9541:a2d4:f0f3:18ff:c868:26ce:e9c4:30a6",
+		gatewayIP:     "9541:a2d4:f0f3:18ff:c868:26ce:e9c4:aaaa",
+		prefixLen:     uint8(64),
+		totalMaskSize: 128,
+	},
+}
+
+func Test_AssignIPToLink(t *testing.T) {
+	ctx := context.Background()
+
+	for _, tt := range defaultAssignIPToLinkTests {
+		t.Run(tt.name, func(st *testing.T) {
+			link1 := newFakeLink(tt.ifStr, 0)
+
+			netlinkAddrAdd = standardNetlinkAddrAdd(tt.allocatedIP, int(tt.prefixLen), tt.totalMaskSize)
+			netlinkRouteAdd = standardNetlinkRouteAdd(tt.gatewayIP, 0, 1)
+
+			if err := assignIPToLink(ctx, tt.ifStr, 10, link1, tt.allocatedIP, tt.gatewayIP, tt.prefixLen, false, 1); err != nil {
+				st.Fatalf("assignIPToLink: %s", err)
+			}
+		})
+	}
+
+}
+
+func Test_AssignIPToLink_No_Gateway(t *testing.T) {
+	ctx := context.Background()
+
+	for _, tt := range defaultAssignIPToLinkTests {
+		t.Run(tt.name, func(st *testing.T) {
+			// remove the gateway IP set for the tests
+			tt.gatewayIP = ""
+			link1 := newFakeLink(tt.ifStr, 0)
+
+			netlinkAddrAdd = standardNetlinkAddrAdd(tt.allocatedIP, int(tt.prefixLen), tt.totalMaskSize)
+			netlinkRouteAdd = standardNetlinkRouteAdd(tt.gatewayIP, 0, 1)
+
+			if err := assignIPToLink(ctx, tt.ifStr, 10, link1, tt.allocatedIP, tt.gatewayIP, tt.prefixLen, false, 1); err != nil {
+				st.Fatalf("assignIPToLink: %s", err)
+			}
+		})
+	}
+
+}
+
+func Test_AssignIPToLink_GatewayOutsideSubnet(t *testing.T) {
+	ctx := context.Background()
+
+	var assignIPToLinkTestsGateway = []assignIPToLinkTest{
```

Comment:
am I seeing this wrong or is this same as `defaultAssignIPToLinkTests`?

Reply:
This one has gateway addresses that are outside the subnet of the IP addresses. I figured it was easier to just make new tests than try to change the gateway address based on if the IP was ipv4 or 6.